packages/api/es/indexer.js

```diff
@@ -344,30 +387,39 @@ async function reingest(g) {
  * @returns {Promise} object with response from the three indexer
  */
 async function handlePayload(event) {
-  let payload;
-  const source = get(event, 'EventSource');
+  const executionModel = new Execution();
+  const granuleModel = new Granule();
+  const pdrModel = new Pdr();
 
-  if (source === 'aws:sns') {
-    payload = get(event, 'Sns.Message');
-    payload = JSON.parse(payload);
-  }
-  else {
-    payload = event;
-  }
+  const payload = event.EventSource === 'aws:sns'
+    ? JSON.parse(event.Sns.Message)
+    : event;
+
+  const executionRecord = await buildExecutionRecord(executionModel, payload);
+  const pdrRecord = buildPdrRecord(pdrModel, payload);
+  const granuleRecords = await buildGranuleRecords(granuleModel, payload);
 
-  let executionPromise;
-  const e = new Execution();
-  if (['failed', 'completed'].includes(payload.meta.status)) {
-    executionPromise = e.updateExecutionFromSns(payload);
+  const recordsCount = flatten([executionRecord, granuleRecords, pdrRecord]).length;
+
+  if (recordsCount > 10 || inTestMode()) {
+    if (executionRecord) await executionModel.create(executionRecord);
+    if (pdrRecord) await pdrModel.create(pdrRecord);
+    await granuleModel.create(granuleRecords);
   }
-  else {
-    executionPromise = e.createExecutionFromSns(payload);
+  else if (recordsCount > 0) {
+    const TransactItems = flatten([
+      buildTransactPut(executionModel.tableName, executionRecord),
+      buildTransactPut(pdrModel.tableName, pdrRecord),
+      granuleRecords.map(buildTransactPut(granuleModel.tableName))
+    ]).filter(isNotNil);
+
+    await dynamodb().transactWriteItems({ TransactItems }).promise();
```

Comment:
I also don't understand what's the benefit to update `execution` `pdr` and `granule` in one transaction.  Do we want to save `execution` records regardless of `granule` records?

Reply:
Granule records reference a PDR record and an execution record.  PDR records also reference an execution record.  This is ensuring that, if writing the execution or PDR to the database fails, you don't end up with granules or PDRs that reference non-existent execution or PDR records.